### ISSUE TITLE
Score-CAM is broken

### ISSUE THREAD

**Issue by miguelCalado** [keisen/tf-keras-vis#75]
I tried to run on a custom model and also ran the examples and it seems that score-cam is broken - takes too much time to compute,  most of the time it fails due to memory issues, and when it does finish, the output is just zeros.

Can you confirm that the problem is just not with me?

Cheers.

**Comment by keisen** [comment]
Hi @miguelCalado .
The issue you're facing was NOT reproduced in my environment.
Could you please submit the code snippet to reproduce the issue?

Thanks!

**Comment by miguelCalado** [comment]
Hi @keisen, 
Thank you for the fast response.

Below I describe the following steps that led me to write this issue:

1. I ran into some problems when running the ```attentions.ipnyb``` notebook. I installed the package via pip and for some reason got an older version (0.5.5 instead of the 0.8.0).  Upgraded the library and it worked fine.

2. The example notebook ```attentions.ipynb``` runs fine just as it is, but when I replace the network from a VGG16 to a ResNet50 the problems that I described above appear. ResNet50 has 2048 filters in the last conv layer, which is much more than the 512 of the VGG16, and thus, it is expected to take longer to run the Score-CAM method. On my P5000 GPU it took 1min 52s. Finally, some of the produced heatmaps are returned as 0 (image below). 

![Screenshot from 2021-09-05 17-36-11](https://user-images.githubusercontent.com/37347167/132134463-26f9b14d-8159-48e7-9959-ce1efe7e1948.png)

When running the Faster-ScoreCAM the issues described above disappear and the returned heatmaps seem to be meaningful. 

**Environment specs**:
- python 3.6
- TensorFlow 2.4.1
- numpy 1.19.5.
- scipy 1.5.4 (important for the resizing)

Cheers.

**Comment by keisen** [comment]
@miguelCalado, Thank you for giving us the details of this issue!

> I installed the package via pip and for some reason got an older version (0.5.5 instead of the 0.8.0). Upgraded the library and it worked fine.

We don't maintain the older version, so please always utilize the latest version.
If , with the latest tf-keras-vis, you face any problem (like this issue) or there is any inconvenience, please feel free to let us know.


> some of the produced heatmaps are returned as 0 (image below).

The reason is that the activation un-merged with `the identity shortcut connection` is visualized.
(Although I believe you've already read it, please see [the paper](https://arxiv.org/pdf/1512.03385.pdf) for details of the `identity` or the `identity shortcut connection`)

Refer to the table shown below that is the output of calling `model.summary()`.

```
...
__________________________________________________________________________________________________
conv5_block2_add (Add)          (None, 7, 7, 2048)   0           conv5_block1_out[0][0]           
                                                                 conv5_block2_3_bn[0][0]          
__________________________________________________________________________________________________
conv5_block2_out (Activation)   (None, 7, 7, 2048)   0           conv5_block2_add[0][0]           
__________________________________________________________________________________________________
conv5_block3_1_conv (Conv2D)    (None, 7, 7, 512)    1049088     conv5_block2_out[0][0]           
__________________________________________________________________________________________________
conv5_block3_1_bn (BatchNormali (None, 7, 7, 512)    2048        conv5_block3_1_conv[0][0]        
__________________________________________________________________________________________________
conv5_block3_1_relu (Activation (None, 7, 7, 512)    0           conv5_block3_1_bn[0][0]          
__________________________________________________________________________________________________
conv5_block3_2_conv (Conv2D)    (None, 7, 7, 512)    2359808     conv5_block3_1_relu[0][0]        
__________________________________________________________________________________________________
conv5_block3_2_bn (BatchNormali (None, 7, 7, 512)    2048        conv5_block3_2_conv[0][0]        
__________________________________________________________________________________________________
conv5_block3_2_relu (Activation (None, 7, 7, 512)    0           conv5_block3_2_bn[0][0]          
__________________________________________________________________________________________________
conv5_block3_3_conv (Conv2D)    (None, 7, 7, 2048)   1050624     conv5_block3_2_relu[0][0]        
__________________________________________________________________________________________________
conv5_block3_3_bn (BatchNormali (None, 7, 7, 2048)   8192        conv5_block3_3_conv[0][0]        
__________________________________________________________________________________________________
conv5_block3_add (Add)          (None, 7, 7, 2048)   0           conv5_block2_out[0][0]           
                                                                 conv5_block3_3_bn[0][0]          
__________________________________________________________________________________________________
conv5_block3_out (Activation)   (None, 7, 7, 2048)   0           conv5_block3_add[0][0]           
__________________________________________________________________________________________________
avg_pool (GlobalAveragePooling2 (None, 2048)         0           conv5_block3_out[0][0]           
__________________________________________________________________________________________________
predictions (Dense)             (None, 1000)         2049000     avg_pool[0][0]                   
==================================================================================================
Total params: 25,636,712
Trainable params: 25,583,592
Non-trainable params: 53,120
__________________________________________________________________________________________________
```

When `penultimate_layer=-1` with Resnet50, tf-keras-vis will find the `conv5_block3_3_conv` layer that is the last convolutional layer in the model. However, as  you see the table above, the activation of the layer is before merging with the activation of the `conv5_block2_out` layer that is the `identity` of `conv5_block3`. I believe that ScoreCAM can't correctly visualize the CAM only with the un-merged activation. (And I think that the reason of that Gradcam and Gradcam++ can visualize CAM correctly even if the `conv5_block3_3_conv` layer is specified is because they utilize the gradients to visualize).

So, to make ScoreCAM work correctly with Resnet, we have to specify the `conv5_block3_add` or `conv5_block3_out` layer as penultimate_layer like below.

```python
cam = scorecam(score, X, penultimate_layer='conv5_block3_out', seek_penultimate_conv_layer=False)
```

 Note that, by default, tf-keras-vis seeks a `Convolutional` layer, so we also have to specify `seek_penultimate_conv_layer=False` to specify the layer other than the Convolutional layer.

> ResNet50 has 2048 filters in the last conv layer, which is much more than the 512 of the VGG16, and thus, it is expected to take longer to run the Score-CAM method. On my P5000 GPU it took 1min 52s.

It was about the same processing time in my environment.
If there is free memory of the GPU, it may be speed-up to specify the `batch_size` option as follows. (Default batch_size=32)

```python
cam = scorecam(score, X, penultimate_layer='conv5_block3_out', seek_penultimate_conv_layer=False, batch_size=128)
```

Of course, I also recommend you to utilize Faster-ScoreCAM as you've already tried.


Thanks!

**Comment by keisen** [comment]
I missed writing one point.

> Finally, some of the produced heatmaps are returned as 0 (image below).

As you may know, in ScoreCAM, CAM is applied Relu function (That's, `np.maximum(cam, 0)` or `tf.keras.backend.relu(cam)`).
So that's why  the results are completely zeros.

Thanks!

**Comment by keisen** [comment]
Hi @miguelCalado 

I'm going to close this issue because I believe that the problem was resolved.
If it has NOT been resolved, please feel free to re-open this issue.

Thanks!
